tests/calldata_cost_test.go

```diff
@@ -0,0 +1,230 @@
+package tests
+
+import (
+	"context"
+	"iter"
+	"math/big"
+	"strings"
+	"testing"
+
+	"github.com/0xsoniclabs/sonic/opera"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCallData_Sonic_GasUsed(t *testing.T) {
+
+	// From https://eips.ethereum.org/EIPS/eip-7623
+	// Gas used before Prague update:
+	// > tx.gasUsed = (
+	// >     21000
+	// >     + STANDARD_TOKEN_COST * tokens_in_calldata
+	// >     + execution_gas_used
+	// >     + isContractCreation * (32000 + INITCODE_WORD_COST * words(calldata))
+	// > )
+	testUsedGas(t,
+		opera.SonicFeatures,
+		func(t *testing.T, data []byte) uint64 {
+			// transaction does not create an account
+			// transaction does not execute code
+			// remaining gas is the intrinsic gas
+			intrinsicGas, err := core.IntrinsicGas(data, nil, nil, false, true, true, true)
+			require.NoError(t, err, "Failed to get the intrinsic gas: ", err)
+			return intrinsicGas
+		})
+}
+
+func TestCallData_Allegro_GasUsed(t *testing.T) {
+
+	// From https://eips.ethereum.org/EIPS/eip-7623
+	// Gas used after Prague update:
+	// > tx.gasUsed = (
+	// >     21000
+	// >     +
+	// >     max(
+	// >         STANDARD_TOKEN_COST * tokens_in_calldata
+	// >         + execution_gas_used
+	// >         + isContractCreation * (32000 + INITCODE_WORD_COST * words(calldata)),
+	// >         TOTAL_COST_FLOOR_PER_TOKEN * tokens_in_calldata
+	// >     )
+	// > )
+	testUsedGas(t,
+		opera.AllegroFeatures,
+		func(t *testing.T, data []byte) uint64 {
+			// transaction does not create an account
+			// transaction does not execute code
+			// remaining gas is the intrinsic gas
+			intrinsicGas, err := core.IntrinsicGas(data, nil, nil, false, true, true, true)
+			require.NoError(t, err, "Failed to get the intrinsic gas: ", err)
+
+			floorGas, err := core.FloorDataGas(data)
+			require.NoError(t, err, "Failed to get the floor gas: ", err)
+
+			return max(intrinsicGas, floorGas)
+		})
+}
```

Comment:
Maybe add a test that makes sure that the produced test data for the transactions result in transactions with different gas requirements in the Sonic and Allegro configuration.